src/usb/uf2/ghostfat.c

```diff
@@ -55,18 +58,22 @@ struct TextFile {
   char const *content;
 };
 
-#define NUM_FAT_BLOCKS UF2_NUM_BLOCKS
+
+//--------------------------------------------------------------------+
+//
+//--------------------------------------------------------------------+
+
+#define NUM_FAT_BLOCKS CFG_UF2_NUM_BLOCKS
```

Comment:
since the flash size of nRF is only 1MB. And I am not the original writer of ghostfat as well. Different board ID won't work, better to be different Family ID. But I don't see the need to do so, we could work on it whenever there is a practical scenario.

Reply:
@henrygab , actually I think you make a very good point for packing multiple uf2, unfortunately uf2 only has family option, and doesn't have board id. Else we could pack all the uf2 of adafruit board into a single uf2. @dhalbert may know more on this.

Reply:
@hathach -- you are right, I meant family, not board ID (updated my comment above).  I am also not the original writer of ghostfat, but I have made significant contributions to it, and am an expert in the FAT file systems.

It might not be a bad idea to The other option is to make the breaking change, where each supported nRF52 board gets a unique UF2 identifier ... which may not be a terrible idea.  After all, each board with different pinouts may require a different bootloader, and a .UF2 designed for the PCA10056 will likely fail to work correctly on a different board.  Therefore, there is a strong reason to have unique family ID for each board, and it fits the idea behind UF2.

If you change to unique family IDs for each board, the bootloader could accept both the "legacy" family ID 0xada52840, and the per-board family ID.  This would allow changes to tooling to slowly bubble up through the tooling, while giving greater safety for those UF2 files authored to be board-specific, and allow non-board-specific UF2 files to explicitly choose to use the legacy family ID.

Reply:
@henrygab yeah, I know you are real FAT expert with the previous PR to ghostfat. I spent tons of time with FAT but still couldn't make most sense out of your last PR :D.  I really like to have ID for each board, unfortunately we have to play along the specs of uf2 https://github.com/microsoft/uf2 The family ID is meant for MCU. If we want to push out for extra field of board ID, we need to get approval from uf2 maintaining team frist. Or there is another way around ? @mmoskal could you help to shed some light on this matter, thank you :).

Reply:
The ghostfat.c uses FAT16, so we could easily increase to 0xffff - 32M. To go higher we would need to change the sectors per cluster - not sure if that's difficult.

You make a very good point about concatenation - I actually didn't think about that when putting in the size limit.

FamilyID is meant as both board-id and MCU-id - for example, binaries generated from MakeCode Arcade would run on all NRF54840 boards provided they have the right bootloader settings (that define pinout, screen type, etc.). OTOH some other binaries may have pin configuration baked in, so they will only work on say 840 Feather. So it actually makes perfect sense for the bootloader to accept MCU-wide familyID and a specific boardID.

Generally, this can thought of as a sub-divsion tree of devices and bootloader accepts everything on one path from root to leaf. In this case I think we have 3 levels (all UF2 file - familyID==0, NRF52840, specific board), but it could be more if needed (eg different revisions of a specific board).

Just make sure to generate them randomly (eg with that CF2 website or your favorite RNG), to minimize risk of collision. We could move it out of README.md into a new file in the UF2 repo to keep track of all these for double-assurance.

Reply:
> FamilyID is meant as both board-id and MCU-id - for example, binaries generated from MakeCode Arcade would run on all NRF54840 boards provided they have the right bootloader settings (that define pinout, screen type, etc.). OTOH some other binaries may have pin configuration baked in, so they will only work on say 840 Feather. So it actually makes perfect sense for the bootloader to accept MCU-wide familyID and a specific boardID.

Thank you very much for answering, yeah the C2F is indeed very useful to allow the same application binary to be able to run on board containing it. That is also why I make it mandatory now for all the board in this PR. And also use its BOATLOADER_ID as verification to prevent incorrect upgrade to another board bootloader. Sorry for the `ADA52840`, I was on time pressure and didn't know the uf2 and its procedure back then. I have done it better with the uf2 bootloader magic of `0xd663823c` generated from your patcher tool. This is used to know if uf2 is containing bootloader or app. Bootloader upgrading is high risk and detection based on address is not trustworthy (app with incorrect linker setting) 

For the Family ID usage as board id, can we use the Adafruit VID/PID board for that purpose. Although it is not randomly generated, it is easily enough to prevent collision because of its publicity ? The bootloader yeah, can be modify to accept `ADA52840` as generic nRF52 binary, and/or VID+PID combo as its specifically built on !! Yahh @dhalbert

Reply:
These are all good ideas. I had a side conversation with @hathach with ideas (some of which already mentioned here):
1. concatenated uf2 with multiple board versions in one
2. generic CircuitPython, etc. with pin config info and other info (e.g. crystal choices) in cf2
3. selecting between translations using cf2, or maybe using filename to select which blocks to load (?? just a random thought)

Reply:
VID/PID sounds like a good idea - the Arcade NRF52840 "board" is really just an 840 feather with a shield (of which a grand total of 1 exists in the world), I wouldn't worry about it too much.

The 0xada52840 isn't so much of a problem either - it's unlikely someone else will use it to mean something different. What I really want to prevent is 0x42424242 and similar since more than one person may such a brilliant idea.

Reply:
Awesome!  It will be good to avoid the accident of putting a .UF2 built for (as an example) a Circuit Playground onto a Feather.  I'm glad the option to allow two family IDs is within scope of UF2 also.

@hathach -- Will you be enabling this in the current PR, or should I keep my eyes open for a future update?

Reply:
awesome, thank you @henrygab for suggesting the idea, and @mmoskal for in-depth information of uf2, and @dhalbert for nice sumup. I have created 2 separated issues regarding #129 and #130 . Since the current PR already made massive changes to the bootloader alone already.  Sorry everyone for getting tagged in multiple places. More great features to looking forward :)